### PR TITLE
JSONStack: Improve performance

### PR DESCRIPTION
**Pull request by obastemur**
1% Acme Air LTO gain. Majority of the time, there is at max 1 item on the stacks and SList too expensive to maintain 1item in/out during the whole JSON.stringify

### CODE REVIEW

File: lib/Runtime/Library/JSONStack.cpp [hunk 1]
```diff
@@ -37,22 +54,68 @@ namespace JSON
     {
         if (bJsObject)
         {
-            return jsObjectStack.Push(data);
+            bool result = true;
```
Comment (digitalinfinity) [inline lib/Runtime/Library/JSONStack.cpp]: Consider refactoring this into an inline function- something like `PushObject(data, jsObjectStack, &STACK_VALUE)/PushObject(data, domObjectStack, &DOM_VALUE)`

File: lib/Runtime/Library/JSONStack.h [hunk 2]
```diff
@@ -15,6 +15,7 @@ namespace JSON
     class JSONStack
     {
     private:
+        Js::Var        tempValues[2];
```
Comment (digitalinfinity) [inline lib/Runtime/Library/JSONStack.h]: Not really your change, but more a question- are these references to Vars supposed to be kept alive? I see that this class uses an arena allocator so curious who keeps it alive

File: lib/Runtime/Library/JSONStack.cpp [hunk 3]
```diff
@@ -24,10 +33,18 @@ namespace JSON
     {
         if (bJsObject)
         {
+            if (STACK_VALUE)
+            {
+                return (data == STACK_VALUE);
```
Comment (kunalspathak) [inline lib/Runtime/Library/JSONStack.cpp]: >return (data == STACK_VALUE); [](start = 16, length = 29)

just to guard future changes to this class, can you add assert at appropriate places to check when STACK_VALUE != nullptr, jsobjectStack is always empty and similarly for dom stack?

File: lib/Runtime/Library/JSONStack.h (hunk 2 above)
Comment (obastemur) [inline lib/Runtime/Library/JSONStack.h]: @digitalinfinity not sure about the answer. I added a destructor to make sure they are nullified.

File: lib/Runtime/Library/JSONStack.cpp (hunk 3 above)
Comment (obastemur) [inline lib/Runtime/Library/JSONStack.cpp]: @kunalspathak  not sure I understood. there is an `if(STACK_VALUE)` why should I check STACK_VALUE != nullptr inside an Assert?

File: lib/Runtime/Library/JSONStack.cpp (hunk 1 above)
Comment (obastemur) [inline lib/Runtime/Library/JSONStack.cpp]: @digitalinfinity This also came to my mind. Good idea. However, all three functions may end up in similar inline function pairs. Current looked more cleaner (whole file is ~100 lines)

**Comment by obastemur** [comment]
@dotnet-bot test Ubuntu shared_ubuntu_linux_test please

**Comment by obastemur** [comment]
@digitalinfinity @kunalspathak Thanks for the review!

File: lib/Runtime/Library/JSONStack.h (hunk 2 above)
Comment (obastemur) [inline lib/Runtime/Library/JSONStack.h]: This is not a problem for JSONStack as it always cleans up the references once stringify is done with inner object/array etc.

File: lib/Runtime/Library/JSONStack.cpp (hunk 3 above)
Comment (kunalspathak) [inline lib/Runtime/Library/JSONStack.cpp]: I meant to say add assert something like
```c++
if(STACK_VALUE) 
{
  assert(jsObjectStack.Count() == 0);
   return (data == STACK_VALUE)
}
```

**Pull request merged**